openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -644,87 +782,402 @@ def _add_to_existing_draft_change_log(
+        if not deps:
+            record_ids_to_hash_digests[record.id] = ''
+            continue
+
+        # If we've gotten this far, it means that this record has dependencies
+        # and does need to get a hash computed for it.
+        records_that_need_hashes.append(record)
+        record_ids_to_live_deps[record.id] = deps
+
+    if backfill:
+        untrusted_record_id_set = None
+    else:
+        untrusted_record_id_set = set(rec.id for rec in records_that_need_hashes)
+
+    for record in records_that_need_hashes:
+        record.dependencies_hash_digest = hash_for_log_record(
+            record,
+            record_ids_to_hash_digests,
+            record_ids_to_live_deps,
+            untrusted_record_id_set,
+        )
+
+    _bulk_update_hashes(record_cls, records_that_need_hashes)
+
+
+def _bulk_update_hashes(model_cls, records):
+    """
+    bulk_update using the model class (PublishLogRecord or DraftChangeLogRecord)
+
+    The only reason this function exists is because mypy 1.18.2 throws an
+    exception in validate_bulk_update() during "make quality" checks otherwise
+    (though curiously enough, not when that same version of mypy is called
+    directly). Given that I'm writing this on the night before the Ulmo release
+    cut, I'm not really interested in tracking down the underlying issue.
+
+    The lack of type annotations on this function is very intentional.
+    """
+    model_cls.objects.bulk_update(records, ['dependencies_hash_digest'])
+
+
+def hash_for_log_record(
+    record: DraftChangeLogRecord | PublishLogRecord,
+    record_ids_to_hash_digests: dict,
+    record_ids_to_live_deps: dict,
+    untrusted_record_id_set: set | None,
+) -> str:
+    """
+    The hash digest for a given change log record.
+
+    Note that this code is a little convoluted because we're working hard to
+    minimize the number of database requests. All the data we really need could
+    be derived from just the record itself, but at a far higher cost.
+
+    When we say "dependency", we mean when one or more PublishableEntities are
+    dependencies of a PublishableEntityVersion. A change in the Draft or
+    Published state of the PublishableEntity causes an implicit change to the
+    PublishableEntityVersion. The common case we have at the moment is when a
+    container type like a Unit has Components as dependencies, i.e. the
+    UnitVersion specifies Component Entities as dependencies.
+
+    This means that the dependencies_hash_digest that summarizes the total state
+    of a PublishableEntity depends on:
+
+    1. The definition of the current draft/published version of that entity.
+    2. The current draft/published versions of all dependencies.
+
+    Which is why it makes sense to capture in a log record, since
+    PublishLogRecords or DraftChagneLogRecords are created whenever one of the
+    above two things changes.
+
+    Here are the possible scenarios, including edge cases:
```

Comment:
[optional] i would reorder this to start with the regular cases and end with the edge cases